subxt/src/rpc/types.rs

```diff
@@ -476,34 +533,101 @@ pub enum FollowEvent<Hash> {
+/// The storage item received as paramter.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StorageQuery<Key> {
+    /// The provided key.
+    pub key: Key,
+    /// The type of the storage query.
+    #[serde(rename = "type")]
+    pub query_type: StorageQueryType,
+}
+
+/// The type of the storage query.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum StorageQueryType {
+    /// Fetch the value of the provided key.
+    Value,
+    /// Fetch the hash of the value of the provided key.
+    Hash,
+    /// Fetch the closest descendant merkle value.
+    ClosestDescendantMerkleValue,
+    /// Fetch the values of all descendants of they provided key.
+    DescendantsValues,
+    /// Fetch the hashes of the values of all descendants of they provided key.
+    DescendantsHashes,
+}
+
+/// The storage result.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StorageResult {
+    /// The hex-encoded key of the result.
+    pub key: String,
+    /// The result of the query.
+    #[serde(flatten)]
+    pub result: StorageResultType,
+}
+
+/// The type of the storage query.
 #[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct ChainHeadResult<T> {
-    /// Result of the method.
-    pub result: T,
+pub enum StorageResultType {
+    /// Fetch the value of the provided key.
+    Value(String),
+    /// Fetch the hash of the value of the provided key.
+    Hash(String),
+    /// Fetch the closest descendant merkle value.
+    ClosestDescendantMerkleValue(String),
 }
 
-/// The event generated by the body and call methods.
+/// The method respose of `chainHead_body`, `chainHead_call` and `chainHead_storage`.
 #[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
 #[serde(rename_all = "camelCase")]
-#[serde(tag = "event")]
-pub enum ChainHeadEvent<T> {
-    /// The request completed successfully.
-    Done(ChainHeadResult<T>),
-    /// The resources requested are inaccessible.
-    ///
-    /// Resubmitting the request later might succeed.
-    Inaccessible(ErrorEvent),
-    /// An error occurred. This is definitive.
-    Error(ErrorEvent),
-    /// The provided subscription ID is stale or invalid.
-    Disjoint,
+#[serde(tag = "result")]
+pub enum MethodResponse {
+    /// The method has started.
+    Started(MethodResponseStarted),
+    /// The RPC server cannot handle the request at the moment.
+    LimitReached,
+}
+
+/// The `started` result of a method.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MethodResponseStarted {
+    /// The operation id of the response.
+    pub operation_id: String,
+    /// The number of items from the back of the `chainHead_storage` that have been discarded.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
```

Comment:
why is serde(default) needed here?  if the field "discarded_items" is missing and it should be deserialized to   Option::None AFAIU.